### PR TITLE
az612: Add joining/down status, fix bugs

### PR DESCRIPTION
**Pull request by jtuple**
Add 'joining' member status to new cluster membership model and
implementation. When a node joins a cluster, it comes in with status
'joining' rather than 'valid'. The claimant then moves a node from 'joining'
to 'valid' after it ensures all cluster members have learned of the new
node joining the cluster. This change guarantees that all 'valid' members
vote on ring ready consensus under various failure scenarios.

Add 'down' member status to new cluster membership model and implementation.
The state is designed to allow a user to mark a down node as 'down' in order
to allow the rest of the cluster to converge. A vote from a 'down' node is
not necessary for ring ready consensus, and therefore a 'down' node's ring
state may become outdated. If a 'down' node gossips to another node that
believes it to be down (such as after coming back online), the other node
tells the 'down' node to rejoin the cluster, therefore making its state
current. Nodes do not gossip to 'down' nodes, and 'down' node ownership is
not changed during a rebalance.

Incorporate minor changes and bug fixes:
-- Fix next merging bug in riak_core_ring:remove_node and model.
-- Fix bug with nodes moving from 'leaving' to 'exiting' while having pending
   indices.
-- Fix negative random seed bug in join/membership model.
-- Change core vnode so that a vnode does not shutdown if a completed handoff
   was to node that is now known to be 'invalid'.
-- Change update_ring to remove tuples from next for all invalid nodes, even
   for completed transfers.
-- Change leave in model to be a local transition like in the implementation.
-- Handle "neither claimant valid" case in reconcile_ring.

### CODE REVIEW

File: src/riak_core_ring_handler.erl [hunk 1]
```diff
@@ -60,11 +60,14 @@ ensure_vnodes_started(Ring) ->
...
             ok;
-        AppMods ->            
+        AppMods ->
             case ensure_vnodes_started(AppMods, Ring, []) of
-                [] -> 
-                    case riak_core_ring:member_status(Ring, node()) of
-                        leaving ->
+                [] ->
+                    Ready = riak_core_ring:ring_ready(Ring),
+                    FutureIndices = riak_core_ring:future_indices(Ring),
```
Comment (jonmeredith) [inline src/riak_core_ring_handler.erl]: Missing an argument to future_indices here - only future_indices/2 is exported.  Crashed while running console status commands.

**Comment by jonmeredith** [comment]
Had a strange issue with leaving a node.

Joined 24-node cluster

```
(dev2@127.0.0.1)17> riak_core_console:member_status().
================================= Membership ==================================
Status     Ring    Pending    Node
-------------------------------------------------------------------------------
... (rest valid)
valid      4.3%      --      'dev23@127.0.0.1'
valid      4.3%      --      'dev24@127.0.0.1'
... (rest valid)
-------------------------------------------------------------------------------
Valid:24 / Leaving:0 / Exiting:0 / Joining:0 / Down:0
```

Downed dev24

Leave on dev23 - that shows up, no marking of dev24 being marked down

```
(dev2@127.0.0.1)18> 09:54:13.402 [info] 'dev23@127.0.0.1' changed from 'valid' to 'leaving'
(dev2@127.0.0.1)18> riak_core_console:member_status().
================================= Membership ==================================
Status     Ring    Pending    Node
-------------------------------------------------------------------------------
leaving    4.3%      --      'dev23@127.0.0.1'
valid      4.3%      --      'dev10@127.0.0.1'
valid      4.3%      --      'dev24@127.0.0.1'
... (all valid)
```

but then sometime later without me issuing any other commands

```
(dev2@127.0.0.1)25> riak_core_console:member_status().
================================= Membership ==================================
Status     Ring    Pending    Node
-------------------------------------------------------------------------------
joining    4.3%      --      'dev23@127.0.0.1'
valid      4.3%      --      'dev10@127.0.0.1'
... (all valid)
valid      4.3%      --      'dev24@127.0.0.1'
... (all valid)
-------------------------------------------------------------------------------
Valid:23 / Leaving:0 / Exiting:0 / Joining:1 / Down:0
ok
```

dev23 is trying to join again.

**Comment by jonmeredith** [comment]
My mistake - I downed 23 instead

```
(dev2@127.0.0.1)21> riak_core:down('dev23@127.0.0.1').
ok
(dev2@127.0.0.1)22> 09:55:01.750 [info] 'dev23@127.0.0.1' changed from 'down' to 'joining'
09:55:01.895 [info] monitor long_gc <0.6455.0> [{initial_call,{erlang,apply,2}},{almost_current_function,{lists,foldl,3}}] [{timeout,
```

File: src/riak_core_gossip.erl [hunk 2]
```diff
@@ -95,7 +98,8 @@ recursive_gossip(Ring) ->
     recursive_gossip(Ring, node()).
 
 random_recursive_gossip(Ring) ->
-    RNode = riak_core_ring:random_node(Ring),
+    Active = riak_core_ring:active_members(Ring),
```
Comment (jonmeredith) [inline src/riak_core_gossip.erl:109]: Maybe node() should be removed from active?  there's no point gossiping to ourselves.

**Comment by jonmeredith** [comment]
Next fun case.  Mass leave of cluster - contained nodes 1-24 minus 19, 21, 23.
dev1 was the claimant.  Told all leaves except dev2 to leave.

dev1 crashed a process before exiting

```
11:51:48.267 [error] gen_server riak_core_gossip terminated with reason: {{function_clause,[{orddict,fetch,[dev1@127.0.0.1,[{dev2@127.0.0.1,[dev2@127.0.0.1,dev2@127.0.0.1]}]]},{riak_core_gossip,recursive_gossip,2},{riak_core_ring_manager,handle_call,3},{gen_server2,handle_msg,7},{proc_lib,init_p_do_apply,3}]},{gen_server2,call,[riak_core_ring_manager,{ring_trans,#Fun<riak_core_gossip.1.124344311>,[{chstate,dev2@127.0.0.1,[{dev10@127.0.0.1,{342,63481859503}},{dev11@127.0.0.1,{425,63481859500}},{dev12@127.0.0.1,{431,63481859507}},{dev13@127.0.0.1,{451,63481859425}}...]...}]}...]}}

```

dev2 is still up, but hasn't become claimant

```
(dev2@127.0.0.1)44> riak_core_console:ring_status().
================================== Claimant ===================================
Claimant:  'dev1@127.0.0.1'
Status:     down
Ring Ready: unknown

============================== Ownership Handoff ==============================
No pending changes.

============================== Unreachable Nodes ==============================
The following nodes are unreachable: ['dev1@127.0.0.1']

WARNING: The cluster state will not converge until all nodes
are up. Once the above nodes come back online, convergence
will continue. If the outages are long-term or permanent, you
can either mark the nodes as down (riak-admin down NODE) or
forcibly remove the nodes from the cluster (riak-admin
force-remove NODE) to allow the remaining nodes to settle.
ok
(dev2@127.0.0.1)45> 
```

All other nodes exited.

**Comment by jonmeredith** [comment]
Retried the mass-exit including claimant - no crashes this time, the original claimant exited, however the last node (dev2) didn't spot it and did not become the claimant.

```
(dev2@127.0.0.1)4> riak_core_console:member_status().
================================= Membership ==================================
Status     Ring    Pending    Node
-------------------------------------------------------------------------------
exiting    0.0%      --      'dev1@127.0.0.1'
valid    100.0%      --      'dev2@127.0.0.1'
-------------------------------------------------------------------------------
Valid:1 / Leaving:0 / Exiting:1 / Joining:0 / Down:0
(dev2@127.0.0.1)5> riak_core_console:ring_status().  
================================== Claimant ===================================
Claimant:  'dev1@127.0.0.1'
Status:     down
Ring Ready: unknown

============================== Ownership Handoff ==============================
No pending changes.

============================== Unreachable Nodes ==============================
The following nodes are unreachable: ['dev1@127.0.0.1']

WARNING: The cluster state will not converge until all nodes
are up. Once the above nodes come back online, convergence
will continue. If the outages are long-term or permanent, you
can either mark the nodes as down (riak-admin down NODE) or
forcibly remove the nodes from the cluster (riak-admin
force-remove NODE) to allow the remaining nodes to settle.

```

**Comment by jonmeredith** [comment]
Next bit of mischief.

rpc:multicall-join a cluster and immediately quit the claimant, something like

```
rpc:multicall(CN, riak_core, join, [node()]). q().
```

Then on another node, mark the claimant node as down.  It didn't start making progress, so I removed the claimant.

```
(dev2@127.0.0.1)9> riak_core_console:ring_status().
================================== Claimant ===================================
Claimant:  'dev1@127.0.0.1'
Status:     down
Ring Ready: unknown

============================== Ownership Handoff ==============================
Owner:      dev1@127.0.0.1
Next Owner: dev12@127.0.0.1

!!! dev1@127.0.0.1 is DOWN !!!

Index: 0
Index: 11417981541647679048466287755595961091061972992
Index: 22835963083295358096932575511191922182123945984
...
-------------------------------------------------------------------------------

============================== Unreachable Nodes ==============================
All nodes are up and reachable

ok
```

Nothing happened, so I removed dev1

```
(dev2@127.0.0.1)17> riak_core:remove('dev1@127.0.0.1').
ok
(dev2@127.0.0.1)11> riak_core_console:ring_status().   
================================== Claimant ===================================
Claimant:  'dev1@127.0.0.1'
Status:     down
Ring Ready: unknown

============================== Ownership Handoff ==============================
Owner:      dev1@127.0.0.1
Next Owner: dev12@127.0.0.1

!!! dev1@127.0.0.1 is DOWN !!!

Index: 0
Index: 11417981541647679048466287755595961091061972992
Index: 22835963083295358096932575511191922182123945984
Index: 34253944624943037145398863266787883273185918976
...

-------------------------------------------------------------------------------

============================== Unreachable Nodes ==============================
The following nodes are unreachable: ['dev1@127.0.0.1']

WARNING: The cluster state will not converge until all nodes
are up. Once the above nodes come back online, convergence
will continue. If the outages are long-term or permanent, you
can either mark the nodes as down (riak-admin down NODE) or
forcibly remove the nodes from the cluster (riak-admin
force-remove NODE) to allow the remaining nodes to settle.
ok

(dev2@127.0.0.1)12> riak_core_console:member_status(). 
================================= Membership ==================================
Status     Ring    Pending    Node
-------------------------------------------------------------------------------
joining    0.0%      0.0%    'dev10@127.0.0.1'
joining    0.0%      0.0%    'dev11@127.0.0.1'
joining    0.0%     50.0%    'dev12@127.0.0.1'
joining    0.0%      0.0%    'dev13@127.0.0.1'
joining    0.0%      0.0%    'dev14@127.0.0.1'
joining    0.0%      0.0%    'dev15@127.0.0.1'
joining    0.0%      0.0%    'dev16@127.0.0.1'
joining    0.0%      0.0%    'dev17@127.0.0.1'
joining    0.0%      0.0%    'dev18@127.0.0.1'
joining    0.0%      0.0%    'dev19@127.0.0.1'
joining    0.0%      0.0%    'dev20@127.0.0.1'
joining    0.0%      0.0%    'dev21@127.0.0.1'
joining    0.0%      0.0%    'dev22@127.0.0.1'
joining    0.0%      0.0%    'dev23@127.0.0.1'
joining    0.0%      0.0%    'dev24@127.0.0.1'
joining    0.0%      0.0%    'dev2@127.0.0.1'
joining    0.0%      0.0%    'dev3@127.0.0.1'
joining    0.0%      0.0%    'dev4@127.0.0.1'
joining    0.0%      0.0%    'dev5@127.0.0.1'
joining    0.0%      0.0%    'dev6@127.0.0.1'
joining    0.0%      0.0%    'dev7@127.0.0.1'
joining    0.0%      0.0%    'dev8@127.0.0.1'
joining    0.0%      0.0%    'dev9@127.0.0.1'
-------------------------------------------------------------------------------
Valid:0 / Leaving:0 / Exiting:0 / Joining:23 / Down:0
ok

```

No progress.

**Comment by jtuple** [comment]
Why do we need Quickcheck when we have Jon?

So, you just hit a case that is known to be undefined in the protocol. To be more specific, the 'joining' status is a weak state, much like 'down'. I added the state to limit the scope of undefined behavior. In short, there is no guarantees about a node in the 'joining' state. However, once a node moves from 'joining' to 'valid', we have very strong guarantees about it's ring consistency and existence as a member in the cluster. Prior to the addition of 'joining', the 'valid' state was basically a weak state -- usually behaves as desired, but would sometimes misbehave under the right conditions. With two states, we now have strong guarantees for 'valid'. Likewise, this is precisely why 'valid' is allowed to claim partitions, but 'joining' is not. A node cannot gain ownership over data until it is in the stronger state.

In short, this is another education / thing to explain / semantics issue. A node isn't a true member of the cluster until it's 'valid', and if you somehow manage to construct a cluster of all 'joining' nodes (as you did), you basically don't have a cluster. Specifically, said cluster will not actually own any data, therefore it should be safe to deconstruct and build again.

In your particular case, 'dev1' is a cluster onto itself. Unless it's given the time to acknowledge other nodes and integrate them into it's view of the world, the other nodes are not part of the cluster.

**Comment by jonmeredith** [comment]
Beware the dreaded self-join

```

(dev1@127.0.0.1)17> riak:join(node()).
ok
(dev1@127.0.0.1)18> riak_core_console:member_status().
================================= Membership ==================================
Status     Ring    Pending    Node
-------------------------------------------------------------------------------
joining  100.0%      --      'dev1@127.0.0.1'
-------------------------------------------------------------------------------
Valid:0 / Leaving:0 / Exiting:0 / Joining:1 / Down:0
```

**Comment by jonmeredith** [comment]
+1 in combination with https://github.com/basho/riak_core/pull/73 and once the join-to-self case is fixed.

Without the vclock change in the p/r I saw gossip overload when the rings did not converge.
